todo_app/app.py

```diff
@@ -17,41 +21,109 @@ def del_DB(db):
     app.config.from_object('todo_app.flask_config.Config')
+    app.config['LOGIN_DISABLED'] = os.getenv('LOGIN_DISABLED')
 
     if db == None:
         db = get_DB()
 
-    @app.route('/', methods=["GET"])
+    #login manager
+    login_manager = LoginManager()
+
+    client = WebApplicationClient(os.getenv('GITHUB_CLIENTID'))
+
+    @login_manager.unauthorized_handler
+    def unauthenticated():
+        authorization_endpoint = 'https://github.com/login/oauth/authorize'
+        request_uri = client.prepare_request_uri(
+            authorization_endpoint
+        )
+        return redirect(request_uri)
+
+    @login_manager.user_loader
+    def load_user(user_id):
+        if user_id in users:
+            return users[user_id] 
+        else:
+            return None
+
+
+    login_manager.init_app(app)
+
+    @app.route('/users/callback', methods=["GET"])
+    def callback_func():
+        token_endpoint='https://github.com/login/oauth/access_token'
+        token_url, headers, body = client.prepare_token_request(
+            token_endpoint,
+            authorization_response=request.url,
+            redirect_url=request.base_url,
+            client_secret=os.getenv('GITHUB_CLIENTSECRET')
+        )
+        token_response = requests.post(
+            token_url,
+            headers=headers,
+            data=body,
+            auth=(os.getenv('GITHUB_CLIENTID'), os.getenv('GITHUB_CLIENTSECRET'))
+        )
+        # Parse the tokens!
+        client.parse_request_body_response(token_response.content.decode())
+
+        # get user details
+        userinfo_endpoint = 'https://api.github.com/user'
+        uri, headers, body = client.add_token(userinfo_endpoint)
+        userinfo_response = requests.get(uri, headers=headers, data=body)
+        if userinfo_response.json().get("login"):
+            login = userinfo_response.json()["login"]
+            unique_id = userinfo_response.json()["id"]
+            email = userinfo_response.json()["email"]
+            avatar = userinfo_response.json()["avatar_url"]
+        else:
+            return "User email not available or not verified by Github.", 400
+        user = User(unique_id,login, email, avatar)
+        if login_user(user):
+            users[str(unique_id)] = user
+            return redirect(url_for("index_get"))
+
+    @app.route('/')
+    @login_required
     def index_get():
         items =  session_items.get_items(db)
-        item_view_model = ViewModel(items)
+        if current_user.get_id() is None:
+            item_view_model = ViewModel(items,User())
+        else: 
+            item_view_model = ViewModel(items,users[str(current_user.get_id())])
+        
         return render_template('index.html',view_model=item_view_model)
 
     @app.route('/delete', methods=["POST"])
+    @login_required
     def delete_item():
         val = request.form.get("deletebutton")
         session_items.delete_item(db,val)
         return redirect(url_for('index_get'))
```

Comment:
There's nothing within this method to actually check the user is a writer - you're relying on the frontend to not show the relevant buttons which is good UX but from a security perspective leaves a big hole; if I know what endpoint to talk to then I can bypass your security. We should definitely add some checks! This might follow a similar pattern to your login function for accounting for logging in being disabled;
```python
def delete_item():
        if (current_user.get_id() != None) and (not current_user.isWriter()):
            return "Permission denied", 401
       ... # carry on and actually do the deletion etc.
```

Obviously we should add this consideration to any endpoint that should check the users role (adding/updating items etc.)